### PR TITLE
Fix const checks inside destructuring expressions

### LINKED ISSUE #182: SyntaxError when destructuring onto instance properties
**Describe the bug**

**Demo:** https://observablehq.com/d/1fb03f01444425d2

Given a cell `foo` with
```javascript
foo = "bar"
```
when defining a cell where the value is referenced as a destructuring default for an instance property:
```javascript
function Test1(opts = {}) {
  ({a: this.a = foo} = opts);
}  
```
Observable will erroneously report
```
SyntaxError: Assignment to constant variable foo
```
To avoid the error, the cell value has to be assigned to an intermediate variable.

For example, the following pattern works:
```javascript
function Test2(opts = {}) {
  const _foo = foo;
  ({a: this.a = _foo} = opts);
}  
```
as does:
```javascript
function Test3(opts = {}, bar = foo) {
  ({a: this.a = bar} = opts);
}  
```

### PR DESCRIPTION
**Pull request by mbostock**
Fixes #182.

The previous logic for ArrayPattern and ObjectPattern was two strict: _every_ identifier inside these nodes was assumed to be an assignment, but in _e.g._ `({foo: value} = …)` and `({foo: value = bar})` only `value` is being assigned to, not `foo` or `bar`. In the new logic, only an ArrayPattern’s elements and an ObjectPattern’s property values are directly tested for assignment; everything else is detected via AssignmentPattern through the outer walk.

### CODE REVIEW

File: src/references.js [hunk 1]
```diff
@@ -170,16 +170,16 @@ export default function findReferences(cell, globals) {
         identifier(node, parents);
         break;
       }
-      case "ArrayPattern":
+      case "ArrayPattern": {
```
Comment (duaneatat) [inline src/references.js:173]: minor nit (feel free to ignore) -- now that the `identifier` function is only used in one `case`, is it worth inlining it? That would remove the shadowing of the `identifier` function, as well as remove the unnecessary mix of `parents` and `nodeParents` (which are the same) in that function.

File: src/references.js [hunk 2]
```diff
@@ -170,16 +170,16 @@ export default function findReferences(cell, globals) {
...
+        for (const element of node.elements) {
+          checkConst(element, parents);
+        }
+        break;
+      }
       case "ObjectPattern": {
-        ancestor(
-          node,
-          {
-            Identifier: identifier,
```
Comment (duaneatat) [inline src/references.js:178]: one minor nit: now that `identifier` is only used in one case, should we remove it and in-line the code? that'd remove the shadowing of `identifier` as well as remove the unnecessary mix of `parents` and `nodeParents` within that function.

**Review by duaneatat — APPROVED** [review]
🎯 lgtm!

File: src/references.js (hunk 2 above)
Comment (mbostock) [reply src/references.js:178]: Oh, good catch, totally. Added! 👍

**Pull request merged**